lib/websocket.go

```diff
@@ -58,14 +102,154 @@ func (h *wsHandler) wsMessageHandler(_ context.Context, e event.Event) error {
-	for k, c := range h.conns {
-		go func(k int, c *websocket.Conn) {
-			err := wsjson.Write(ctx, c, evt)
-			if err != nil {
-				log.Errorf("connection %d: wsjson write error: %s", k, err)
-			}
-		}(k, c)
+	profileIDString := e.ProfileID
+	if profileIDString == "" {
+		log.Debugf("Event with SessionID %q has no scope. Not sending event over websocket.", e.SessionID)
+		return nil
+	}
+	connID, ok := h.subscriptions[profileIDString]
+	if !ok {
+		return fmt.Errorf("no websocket connection ID found for profile %q", profileIDString)
+	}
+	c, ok := h.conns[connID]
+	if !ok {
+		h.unsubscribeConn(profileIDString)
+		return fmt.Errorf("no websocket connection found for connection ID %q, profile %q", connID, profileIDString)
+	}
+	log.Debugf("sending event %q to websocket conns %q", e.Type, profileIDString)
+	err := wsjson.Write(ctx, c.conn, evt)
+	if err != nil {
+		log.Errorf("connection %q: wsjson write error: %s", profileIDString, err)
+	}
+	return nil
+}
+
+// subscribeConn authenticates the given token and adds the connID to the map
+// of "subscribed" connections
+func (h *wsHandler) subscribeConn(connID, tokenString string) error {
+	ctx := context.TODO()
+	tok, err := token.ParseAuthToken(ctx, tokenString, h.keystore)
+	if err != nil {
+		return err
 	}
+
+	claims, ok := tok.Claims.(*token.Claims)
+	if !ok || claims.Subject == "" {
+		return fmt.Errorf("cannot get profile.ID from token")
+	}
+	// TODO(b5): at this point we have a valid signature of a profileID string
+	// but no proof that this profile is owned by the key that signed the
+	// token. We either need ProfileID == KeyID, or we need a UCAN. we need to
+	// check for those, ideally in a method within the profile package that
+	// abstracts over profile & key agreement
+
+	h.connsLock.Lock()
+	c, ok := h.conns[connID]
+	if !ok {
+		return fmt.Errorf("no connection for connection ID %q found", connID)
+	}
+	c.profileID = claims.Subject
+	h.connsLock.Unlock()
+
+	h.subLock.Lock()
+	defer h.subLock.Unlock()
+	h.subscriptions[claims.Subject] = connID
 	return nil
 }
+
+// unsubscribeConn remove the profileID and connID from the map of "subscribed"
+// connections
+func (h *wsHandler) unsubscribeConn(profileID string) {
+	h.subLock.Lock()
+	defer h.subLock.Unlock()
+	delete(h.subscriptions, profileID)
+}
+
+// removeConn removes the conn from the map of connections and subscriptions
+// closing the connection if needed
+func (h *wsHandler) removeConn(connID string) {
+	c, ok := h.conns[connID]
+	if !ok {
+		return
+	}
+	defer func() {
+		c.conn.Close(websocket.StatusNormalClosure, "pruning connection")
+	}()
+	if c.profileID != "" {
+		h.unsubscribeConn(c.profileID)
+	}
+	h.connsLock.Lock()
+	defer h.connsLock.Unlock()
+	delete(h.conns, connID)
+}
+
+// read listens to the given connection, handling any messages that come through
+// stops listening if it encounters any error
+func (h *wsHandler) read(id string) error {
+	ctx := context.Background()
+	msg := &message{}
+	var err error
```

Comment:
I'm a fan of using named return values to avoid needing lines like this, so long as you always explicitly declare returns (eg: `return err`, never `return`)